packages/ses/src/module-link.js

```diff
@@ -0,0 +1,62 @@
+import { makeModuleInstance } from './module-instance.js';
+
+const { entries } = Object;
+// q, as in quote, for quoting strings in error messages.
+const q = JSON.stringify;
+
+// `link` creates `ModuleInstances` and `ModuleNamespaces` for a module and its
+// transitive dependencies and connects their imports and exports.
+// After linking, the resulting working set is ready to be executed.
+// The linker only concerns itself with module namespaces that are objects with
+// property descriptors for their exports, which the Compartment proxies with
+// the actual `ModuleNamespace`.
+export const link = (
+  compartment,
+  compartmentPrivateFields,
+  moduleSpecifier,
+) => {
+  const { moduleRecords, instances } = compartmentPrivateFields.get(
+    compartment,
+  );
+
+  // Memoize.
+  if (instances.has(moduleSpecifier)) {
+    return instances.get(moduleSpecifier);
+  }
+
+  const moduleRecord = moduleRecords.get(moduleSpecifier);
+  if (moduleRecord == null) {
+    throw new ReferenceError(`Missing link to module ${q(moduleSpecifier)}`);
+  }
+
+  // The module compartment record graph can refer through an alias to a record
+  // from a different compartment.
+  // We need to evaluate this module in the context of its own global object.
+  const { globalObject } = compartmentPrivateFields.get(
+    moduleRecord.compartment,
+  );
+
+  const importedInstances = new Map();
+  const moduleInstance = makeModuleInstance(
+    moduleRecord,
+    importedInstances,
+    globalObject,
+  );
+
+  // Memoize.
+  instances.set(moduleRecord.moduleSpecifier, moduleInstance);
```

Comment:
AFAICT this is a bug. We should memoize here instead on `moduleSpecifier`.

(As you suggested verbally) it would be good to first write a test that demonstrates the bug.

Reply:
I added a number of tests trying to tease out a bug with a diamond dependency, both between compartments and in the same compartment, but wasn’t able effect a bad behavior. But, as I agree this looks wrong, but I addressed this in the next commit.